consensus/propagation/reactor.go

```diff
@@ -299,14 +313,160 @@ func chunkIndexes(totalSize, chunkSize int) [][2]int {
+		if len(reqs) >= reqLimit {
+			// TODO unify the types for the indexes and similar
+			hc.RemoveIndex(uint32(partIndex))
+			// mark the part as fully requested.
+			fullReqs.SetIndex(partIndex, true)
+		}
+		// don't request the part from this peer if we've already requested it
+		// from them.
+		for _, p := range reqs {
+			// p == peer means we have already requested the part from this peer.
+			if p == peer {
+				hc.RemoveIndex(uint32(partIndex))
+			}
+		}
+	}
+
+	// todo(evan): check that this is legit. we can also exit early if we have
+	// all of the data already
+	if hc.IsEmpty() {
+		return
+	}
+
+	// send a want back to the sender of the haves with the wants we
+	e := p2p.Envelope{
+		ChannelID: WantChannel,
+		Message: &propproto.WantParts{
+			Height: height,
+			Round:  round,
+			Parts:  *hc.ToBitArray().ToProto(),
+		},
+	}
+
+	if !p2p.SendEnvelopeShim(p.peer, e, blockProp.Logger) { //nolint:staticcheck
+		blockProp.Logger.Error("failed to send part state", "peer", peer, "height", height, "round", round)
+		return
+	}
+
+	schema.WriteBlockPartState(
+		blockProp.traceClient,
+		height,
+		round,
+		hc.GetTrueIndices(),
+		false,
+		string(peer),
+		schema.Haves,
+	)
+
+	// keep track of the parts that this node has requested.
+	// TODO check if we need to persist the have parts or just their bitarray
+	p.SetRequests(height, round, hc.ToBitArray())
+	blockProp.broadcastHaves(hc, peer)
+}
+
+// todo(evan): refactor to not iterate so often and just store which peers
+func (blockProp *Reactor) countRequests(height int64, round int32, part int) []p2p.ID {
+	peers := make([]p2p.ID, 0)
+	for _, peer := range blockProp.getPeers() {
+		reqs, has := peer.GetRequests(height, round)
+		if has {
+			index := reqs.GetIndex(part)
+			if index {
+				peers = append(peers, peer.peer.ID())
+			}
+		}
+	}
+	return peers
 }
 
 // broadcastHaves gossips the provided have msg to all peers except to the
 // original sender. This should only be called upon receiving a new have for the
 // first time.
-//
-//nolint:unused
 func (blockProp *Reactor) broadcastHaves(haves *proptypes.HaveParts, from p2p.ID) {
+	e := p2p.Envelope{
+		ChannelID: DataChannel,
+		Message: &propproto.HaveParts{
+			Height: haves.Height,
+			Round:  haves.Round,
+			Parts:  haves.ToProto().Parts,
+		},
+	}
+	for _, peer := range blockProp.getPeers() {
+		if peer.peer.ID() == from {
+			continue
+		}
+
+		// skip sending anything to this peer if they already have all the
+		// parts.
+		ph, has := peer.GetHaves(haves.Height, haves.Round)
+		if has {
+			havesCopy := haves.Copy()
+			havesCopy.Sub(ph.ToBitArray())
+			if havesCopy.IsEmpty() {
+				continue
+			}
+		}
+
+		// todo(evan): don't rely strictly on try, however since we're using
+		// pull based gossip, this isn't as big as a deal since if someone asks
+		// for data, they must already have the proposal.
+		if p2p.SendEnvelopeShim(peer.peer, e, blockProp.Logger) { //nolint:staticcheck
```

Comment:
one of the downsides here is that we could end up blocking for all peers if one peer's send queue is full. we likely need to use try and log